numpy/_core/src/umath/string_fastsearch.h

```diff
@@ -0,0 +1,744 @@
+#ifndef _NPY_CORE_SRC_UMATH_STRING_FASTSEARCH_H_
+#define _NPY_CORE_SRC_UMATH_STRING_FASTSEARCH_H_
+
+/* stringlib: fastsearch implementation taken from CPython */
+
+#include <Python.h>
+#include <limits.h>
+#include <string.h>
+#include <wchar.h>
+
+#include <type_traits>
+
+
+/* fast search/count implementation, based on a mix between boyer-
+   moore and horspool, with a few more bells and whistles on the top.
+   for some more background, see:
+   https://web.archive.org/web/20201107074620/http://effbot.org/zone/stringlib.htm */
+
+/* note: fastsearch may access s[n], which isn't a problem when using
```

Comment:
Hmmm, not sure! Even though NumPy strings are not null-terminated, the most this can do is read 4 bytes past the data buffer of the array, which is okay, no?

Reply:
No it isn't.

Reply:
I've fixed this in https://github.com/numpy/numpy/pull/24868/commits/971ebd46e0d47f954954b637ddd91698b6279244, but, unfortunately, I don't really know how to test it, so I've not added any tests.

Reply:
If you can, create an array completely filled with something where this happens.  Best case, it would do nonsense (because it is not null terminated but gets the beginning of the next string).  Worst case, the error will show up if you test with a sanitizer/valgrind on the very last element in the array.